### PR TITLE
Modified how logger is provided to napps

### PR DESCRIPTION
**Pull request by Ktmi**
This resolves issue #236, by making it so that importing the logger provided through `kytos.core.log` is the appropriate child logger for the given napp. Rather than providing a singleton facade which negotiates the appropriate logger at run time, the new implementation will retrieve the appropriate module at import through the use of a property decorated function.

# Notes

It should be noted that this process is not guaranteed to result in the same behaviour as before. I have yet to see an occasion where a log entry seems to come from the incorrect logger, but it may be a possibility.

### CODE REVIEW

File: kytos/core/__init__.py [hunk 1]
```diff
@@ -18,4 +20,5 @@
 )
 
 # Kept lowercase to be more user friendly.
-log = NAppLog()  # pylint: disable=invalid-name
+log = property(lambda self: self.get_napp_logger()) \
```
[the remark this answers is not in the document]
Comment (viniarck) [reply kytos/core/__init__.py:23]: Also, to clarify and align some requirements, regarding the ultimate log goal, our team also want to make sure that logs can be used on hot paths, if we have an intermediary implementation that performs well enough and don't add that much latency (assuming a test with 100+ req / sec, so please later on make sure to also run some benchmarks, if you need help with the tolling the our team typically uses let me know) we can keep using it it for in the meantime and for this upcoming release until we figure out or can afford to integrate a complete asynchronous logger, potentially with `QueueHandler` and `QueueListerner` as mentioned on this comment https://github.com/kytos-ng/kytos/issues/236#issuecomment-116566961 (that will require more research and assessment to see if can be compatible with the existing logging.ini options that the platform provides).

**Comment by Ktmi** [comment]
@viniarck So I created the following program for benchmarking my change, as well as benchmarking different handlers for logging.

log_bench.py

```python
import inspect
from queue import Queue
import re
from logging import getLogger, Formatter, StreamHandler, INFO, handlers
import datetime

#: Detect NApp ID from filename
NAPP_ID_RE = re.compile(r'.*napps/(.*?)/(.*?)/')


def _detect_napp_id():
	"""Get the last called NApp in caller's stack.

	We use the last innermost NApp because a NApp *A* may call a NApp *B* and,
	when *B* uses the logger, the logger's name should be *B*.

	Returns:
		str, None: NApp ID or None if no NApp is found in the caller's stack.

	"""
	for frame in inspect.stack():
		if not frame.filename == __file__:
			match = NAPP_ID_RE.match(frame.filename)
			if match:
				return '/'.join(match.groups())
	return None

def get_napp_logger(logger):
	"""Detect NApp ID and get its logger."""
	napp_id = _detect_napp_id()

	# if napp_id is detected, get the napp logger.
	if napp_id:
		logger = logger.getChild(napp_id)

	return logger

def provider1():
	base_logger = getLogger("kytos.napps1")
	base_logger.setLevel(INFO)
	class NAppLog:
		"""High-level logger for NApp devs.

		From NApp dev's point of view:
		- No need for instantiation
		- Logger name is automatically assigned

		Redirect all calls to a logger with the correct name (NApp ID).

		The appropriate logger is a logging.Logger with NApp ID as its name. If no
		NApp is found, use the root logger.

		As any NApp can use this logger, its name is detected in every call by
		inspecting the caller's stack. If no NApp is found, use the root logger.
		"""

		def __getattribute__(self, name):
			"""Detect NApp ID and use its logger."""
			return get_napp_logger(base_logger).__getattribute__(name)
	
	return NAppLog()

def provider2():
	base_logger = getLogger("kytos.napps2")
	base_logger.setLevel(INFO)
	return get_napp_logger(base_logger)

providers = {
	'old_provider': provider1,
	'new_provider': provider2
}

fmt = Formatter('%(asctime)s - %(levelname)s [%(name)s] (%(threadName)s) %(message)s')

file_handler = handlers.RotatingFileHandler('kytos.log')
file_handler.setFormatter(fmt)
file_handler.setLevel(INFO)

console_handler = StreamHandler()
console_handler.setFormatter(fmt)
console_handler.setLevel(INFO)

file_queue = Queue()
file_queue_listener = handlers.QueueListener(file_queue, file_handler, True)
file_queue_listener.start()
file_queue_handler = handlers.QueueHandler(file_queue)

console_queue = Queue()
console_queue_listener = handlers.QueueListener(console_queue, console_handler, True)
console_queue_listener.start()
console_queue_handler = handlers.QueueHandler(console_queue)

handles = {
	'file_handler': file_handler,
	'console_handler': console_handler,
	'file_queue_handler': file_queue_handler,
	'console_queue_handler': console_queue_handler,
}

def test_executor(provider, handler):
	test1 = datetime.datetime.now()
	log = provider()
	log.addHandler(handler)
	test2 = datetime.datetime.now()
	n_messages = 10000
	for i in range(n_messages):
		log.info(f'Msg: {i}')
	test3 = datetime.datetime.now()
	log.removeHandler(handler)
	return (test2 - test1, test3 - test2)
	

def main():
	for provider_name, provider in providers.items():
		for handler_name, handler in handles.items():
			init_time, run_time = test_executor(provider, handler)
			print(f'{provider_name}, {handler_name}, {init_time}, {run_time}')

main()
```

This program produces 10,000 records per a logger handler pair. It times how long it takes to initialize the logger, as well as how long it takes to process the 10,000 records. To execute this program, I ran it with the following command:

```bash
python log_bench.py > results.csv
```

The produced results were:
```csv
old_provider, file_handler, 0:00:00.005044, 0:00:02.083217
old_provider, console_handler, 0:00:00.000171, 0:00:15.878804
old_provider, file_queue_handler, 0:00:00.000184, 0:00:01.799022
old_provider, console_queue_handler, 0:00:00.000177, 0:00:01.761531
new_provider, file_handler, 0:00:00.000163, 0:00:00.474181
new_provider, console_handler, 0:00:00.000240, 0:00:03.941527
new_provider, file_queue_handler, 0:00:00.000347, 0:00:00.258122
new_provider, console_queue_handler, 0:00:00.000207, 0:00:00.246152
```

Logging with the new logging provider was significantly faster than the old one, about 4x faster in logging directly to console and directly to file.

Logging with queues was overall faster than logging directly to the file or console, with it taking only ~0.25 seconds for 10,000 records in comparison to logging to file which took ~0.47 seconds for 10,000 records.

**Comment by viniarck** [comment]
> 
> Logging with the new logging provider was significantly faster than the old one, about 4x faster in logging directly to console and directly to file.
> 
> Logging with queues was overall faster than logging directly to the file or console, with it taking only ~0.25 seconds for 10,000 records in comparison to logging to file which took ~0.47 seconds for 10,000 records.

@Ktmi nicely done, significant improvements in this initial benchmark compared to what we have and in this last commit the log name was as expected. David, could you also fix the linter error? Also, could you add a line on the `CHANGELOG.rst`, it would be great to mention the improvements on the current `log`. 

@Ktmi, I checkout this PR locally and I was also exploring a `flow_manager` API endpoint that has a log.info statement in a critical path since with APM is easy to get additional statistical information, and it was also observed that in one of the samples under the 95 percentile a `log.info` can sometimes contribute to the latency significantly, 11ms, compared to the database call for instance 2.8ms, overall though, it seems that there are fewer of these samples: 

2022-07-19-164927_1647x770_scrot

and, for the record, another sample under the 50 percentile it was great, the `log.info` took roughly 320 us:

2022-07-19-164301_1643x779_scrot

I ran these stress tests with [`vegeta`](https://github.com/tsenart/vegeta) by the way:

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": false, "flows": [ { "priority": 10, "match": { "in_port": 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/json"]}}' | vegeta attack -format=json -rate 50/1s -
duration=20s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         1000, 50.05, 50.03
Duration      [total, attack, wait]             19.989s, 19.98s, 8.268ms
Latencies     [min, mean, 50, 90, 95, 99, max]  5.501ms, 15.649ms, 10.278ms, 35.621ms, 45.979ms, 63.295ms, 92.038ms
Bytes In      [total, mean]                     37000, 37.00
Bytes Out     [total, mean]                     123000, 123.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:1000
Error Set:
```

It was great to know that in your bench the queue implementation showed a bit better results, which we can explore in a next iteration, I'll also take an opportunity to try to use the `QueueHandler` I'd expect that it would probably be able to minimize these latency instability.

**Comment by viniarck** [comment]
I ran another stress test using `QueueHandler` and `QueueListener`, overall, the 95 percentile was pretty much the same, but the max latency was slightly lower:

```
❯ jq -ncM '{method: "POST", url: "http://localhost:8181/api/kytos/flow_manager/v2/flows/00:00:00:00:00:00:00:01", body: { "force": false, "flows": [ { "priority": 10, "match": { "in_port": 1, "dl_vlan": 100 }, "actions": [ { "action_type": "output", "port": 1 } ] } ] } | @base64, header: {"Content-Type": ["application/json"]}}' | vegeta attack -format=json -rate 50/1s -
duration=20s | tee results.bin | vegeta report
Requests      [total, rate, throughput]         1000, 50.05, 50.01
Duration      [total, attack, wait]             19.996s, 19.98s, 15.544ms
Latencies     [min, mean, 50, 90, 95, 99, max]  5.546ms, 15.109ms, 10.948ms, 30.644ms, 44.18ms, 64.865ms, 85.668ms
Bytes In      [total, mean]                     37000, 37.00
Bytes Out     [total, mean]                     123000, 123.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      202:1000
Error Set:
```

Under the 95 percentile, there was a log.info statement taking time 6 ms+ than a DB call, so that implies that even with this solution, `log` entries will have to be used cautiously on hot paths (It might be worth doing more research here in the future to see if there's another queued solutions that could enhance this):

2022-07-19-180154_1647x797_scrot

In the bright side though,  when using `log.debug` that used to be a problem, it isn't anymore, so the queued handler is respecting the level correctly without adding additional latency:

2022-07-19-180803_1650x775_scrot 

FYI @Ktmi @italovalcy ^ 

Here's the patched code that I was using for the record:

```
diff --git a/main.py b/main.py
index 3b38f1a..44e5607 100644
--- a/main.py
+++ b/main.py
@@ -8,6 +8,7 @@ from enum import Enum
 from threading import Lock
 
 from flask import jsonify, request
+from kytos.core.apm import execution_context
 from napps.kytos.flow_manager.match import match_flow
 from napps.kytos.of_core.flow import FlowFactory
 from napps.kytos.of_core.msg_prios import of_msg_prio
@@ -22,7 +23,7 @@ from werkzeug.exceptions import (
     UnsupportedMediaType,
 )
 
-from kytos.core import KytosEvent, KytosNApp, log, rest
+from kytos.core import KytosEvent, KytosNApp, rest
 from kytos.core.helpers import listen_to, now
 
 from .barrier_request import new_barrier_request
@@ -47,6 +48,33 @@ from .utils import (
     is_ignored,
 )
 
+from queue import Queue
+from logging import getLogger, Formatter, StreamHandler, INFO, handlers
+
+fmt = Formatter('%(asctime)s - %(levelname)s [%(name)s] (%(threadName)s) %(message)s')
+
+file_handler = handlers.RotatingFileHandler('kytosd.log')
+file_handler.setFormatter(fmt)
+file_handler.setLevel(INFO)
+
+console_handler = StreamHandler()
+console_handler.setFormatter(fmt)
+console_handler.setLevel(INFO)
+
+file_queue = Queue(-1)
+file_queue_listener = handlers.QueueListener(file_queue, file_handler, True)
+file_queue_listener.start()
+file_queue_handler = handlers.QueueHandler(file_queue)
+
+console_queue = Queue(-1)
+console_queue_listener = handlers.QueueListener(console_queue, console_handler, True)
+console_queue_listener.start()
+console_queue_handler = handlers.QueueHandler(console_queue)
+
+log = getLogger("flow_manager_queued")
+log.addHandler(file_queue_handler)
+log.addHandler(console_queue_handler)
+
 
 class FlowEntryState(Enum):
     """Enum for stored Flow Entry states."""
@@ -533,10 +561,15 @@ class Main(KytosNApp):
                 raise BadRequest(result)
 
         force = bool(flows_dict.get("force", False))
-        log.info(
+        transaction = execution_context.get_transaction()
+        if transaction:
+            transaction.begin_span("log", "custom")
+        log.debug(
             f"Send FlowMod from request dpid: {dpid}, command: {command}, "
             f"force: {force}, flows_dict: {flows_dict}"
         )
+        if transaction:
+            transaction.end_span()
         try:
             if not dpid:
                 self._install_flows(
diff --git a/settings.py b/settings.py
index f482b7b..0f49b7e 100644
--- a/settings.py
+++ b/settings.py
@@ -8,7 +8,7 @@ ENABLE_BARRIER_REQUEST = True
 # List of flows ignored by the consistency check
 # To filter by a cookie or `table_id` use [value]
 # To filter by a cookie or `table_id` range [(value1, value2)]
-CONSISTENCY_COOKIE_IGNORED_RANGE = []
+CONSISTENCY_COOKIE_IGNORED_RANGE = [(0xdd00000000000000, 0xdd99999999999999)]
 CONSISTENCY_TABLE_ID_IGNORED_RANGE = []
 
 # Retries options for `kytos/core.openflow.connection.error`
```

**Review by viniarck — APPROVED** [review]
LGTM, well done, @Ktmi. Could you map for us a new issue/enhancement to continue to explore/research the points we've discussed in the last comments? Also, if you could merge `master` into your branch again, it's behind upstream head. 

Thanks.

**Comment by viniarck** [comment]
@Ktmi I went ahead and merged `master` into your branch, and I'll go ahead and merge this PR into `master`. Appreciated your contribution, great to have these core log improvements.

**Pull request merged**